### DISCUSSION TITLE
missing exception templates and probably SSL stuff when going from on-server to morbo devel

### DISCUSSION THREAD

**Discussion by bkerin** [mojolicious/mojo#1945] [Q&A]
I'm trying to use morbo again for some speed after devel on server for a while and I've hit a couple issues:

one is exception templates aren't found, morbo says: [2022-06-21 14:53:05.34259] [14320] [debug] Template "exception.development.html.ep" not found

the other is the Mojo::UserAgent that I've been using successfully from the server doesn't work for SSL reasons, it says: SSL connect attempt failed error:14090086:SSL routines:SSL3_GET_SERVER_CERTIFICATE:certificate verify failed.  I suspect the user agent isn't finding the cert lists that is uses when run from CGI script on server.

It seems like these issues are likely related.  I'm using morbo directly on the script in my source dir 'morbo web_client.cgi' (with some code to pick local files for secret, test db etc.), but it seems like it isn't picking up some global Mojolicious context that the CGI script gets, is there some way to tell morbo where it should be getting that stuff?

**Comment by daleif** [comment]
Did you create `exception.development.html.ep`? It is quite normal for `morbo` to report that an exception template was not found and then use something generic instead. 

What exactly are you using here? `morbo` would normally run `script/app_name`. How is `.cgi` related here?

**Reply by kraih** [reply]
The fact that it's a `[debug]` message also tells me that the Mojolicious version is not up to date. It should be an even lower priority `[trace]` message.

**Comment by bkerin** [comment]
On Wed, Jun 22, 2022 at 7:00 AM Sebastian Riedel ***@***.***>
wrote:

> The fact that it's a [debug] message also tells me that the Mojolicious
> version is not up to date. It should be an even lower priority [trace]
> message.
>
Ok update to latest Mojolicious fixed the weird way the template looked
(missing template trace messages still appear as expected).

The certificate verify failed problem remains on my (older) laptop:

  * Mojo::UserAgent query (of weather.visualcrossing.com) in an isolated
script on my laptop fails
  * Mojo::UserAgent query (of weather.visualcrossing.com) in an isolated
script on the server works
  * Mojo::UserAgent query of google.com on my laptop works

So it looks like the problem may be that there really isn't a trust path on
my laptop given the certs Mojo::UserAgent is finding.
Any additional clues on where to look next to diagnose/fix this would be
greatly appreciated.

Britton

**Comment by bkerin** [comment]
Looks like the problem was an ancient cert on my old laptop.  I de-trusted
it and then UserAgent found the same new cert as on the server and worked.

The odd part from the Mojolicious perspective was the way to debug this
which involved

  IO::Socket::SSL::set_defaults(SSL_verify_callback => sub { say
"Authority: $_[2]" and return $_[0] });

and then you guess the associated pem file which is causing the problem
from the output.  I guess maybe there's no way for the ->ca method to list
the actual file involved but it would be nice if things were that direct.

Britton

**Reply by kraih** [reply]
The error messages come from OpenSSL.

**Comment by bkerin** [comment]
On Wed, Jun 22, 2022 at 11:26 AM Sebastian Riedel ***@***.***>
wrote:

> The error messages come from OpenSSL.
>

I know but especially since UserAgen has ->ca() it would be nice if the
actual file causing the error could get mentioned somehow.  ->ca() without
args seems a getter method but returns undef by default.  All probably
perfectly reasonable and I realize how tough it can be to get this sort of
fail propagated up in any intelligible way, but we can wish.  I'd hazard a
guess that cert fail is a somewhat common source of error at the moment as
http fades away.

Britton
